fcomm/src/bin/fcomm.rs

```diff
@@ -490,11 +496,87 @@ fn main() -> Result<(), Error> {
         .filter_level(cli.verbose.log_level_filter())
         .init();
 
+    // cache lives as long as the process
+    let cache = Box::leak(Box::new(cache::PublicParams {
+        file_store: Path::new(&cache::dir("lurk")).join("pubparams.json"),
+        data: HashMap::<usize, PublicParams>::new(),
+    }));
+
+    cache.read().ok();
+
     match &cli.command {
         Command::Commit(c) => c.commit(cli.limit),
-        Command::Open(o) => o.open(o.chain, cli.limit, cli.eval_input, o.quote_input),
+        Command::Open(o) => o.open(o.chain, cli.limit, cli.eval_input, o.quote_input, cache),
         Command::Eval(e) => e.eval(cli.limit),
-        Command::Prove(p) => p.prove(cli.limit),
-        Command::Verify(v) => v.verify(cli.error),
+        Command::Prove(p) => p.prove(cli.limit, cache),
+        Command::Verify(v) => v.verify(cli.error, cache),
+    }
+}
+
+mod cache {
+    use fcomm::error::Error;
+    use lurk::proof::nova;
+    use std::boxed::Box;
+    use std::collections::HashMap;
+    use std::env;
+    use std::hash::Hash;
+    use std::path::PathBuf;
+
+    pub struct Cache<
+        K: Eq + Hash + serde::Serialize + serde::de::DeserializeOwned,
+        V: serde::Serialize + serde::de::DeserializeOwned,
+    > {
+        pub data: HashMap<K, V>,
+        pub file_store: PathBuf,
+    }
+
+    pub type PublicParams<'a> = Cache<usize, nova::PublicParams<'a>>;
+
+    impl<
+            'a,
+            K: Eq + Hash + serde::Serialize + serde::de::DeserializeOwned,
+            V: serde::Serialize + serde::de::DeserializeOwned,
+        > Cache<K, V>
+    {
+        pub fn read(&mut self) -> Result<(), Error> {
+            let serialization = std::fs::read(self.file_store.clone())?;
+            self.data = serde_json::from_slice(serialization.as_slice())?;
+            Ok(())
+        }
+
+        pub fn write(&mut self) -> Result<(), Error> {
+            std::fs::create_dir_all(self.file_store.as_path().parent().unwrap())?;
+            let db = serde_json::to_string(&self.data)?;
+            std::fs::write(self.file_store.clone(), db)?;
+            Ok(())
+        }
+    }
+    const PREFIX: &str = "LURK";
+
+    /// All cache files and directories paths should be constructed using this function,
+    /// which its base directory from the LURK_CACHE_DIR env var, and defaults to /var/tmp.
+    /// Note that LURK_CACHE_DIR is not a first class setting and can only be set by env var.
+    pub fn dir(s: &str) -> String {
+        let cache_var = format!("{}_CACHE_DIR", PREFIX);
+        let mut cache_name = env::var(cache_var).unwrap_or_else(|_| "/var/tmp/".to_string());
+        cache_name.push_str(s);
+        cache_name
+    }
+
+    pub fn public_params(
```

Comment:
I thought of trying to make this a generalized function that caches whatever function you pass it, rather than just `public_params`. It gets hairy though, not knowing how many inputs the function has or whether they're mappable to json map keys. So I decided that is premature generalization. I'm open to arguments to the contrary though.

Reply:
Yes, I would not do that, for multiple reasons.